### PR TITLE
More Checking on `technology_interconnections`

### PR DESCRIPTION
**Pull request by johnjasa**
# More Checking on `technology_interconnections`

This PR adds structured topology validation for `technology_interconnections` in `H2IntegrateModel`, raising descriptive errors for common configuration mistakes that previously went undetected until OpenMDAO setup or silently produced wrong results.

Summary of changes:
1. Add `_validate_technology_interconnections` method with three topology checks.
2. Extend `tech_control_classifiers` population to always run (was SLC-only).
3. Call the new validation in `setup()` before the existing commodity check.
4. Add integration tests covering all new error cases.

## Section 1: Type of Contribution
- [x] Feature Enhancement
  - [ ] Framework
  - [ ] New Model
  - [ ] Updated Model
  - [x] Tools/Utilities
  - [ ] Other (please describe):
- [ ] Bug Fix
- [ ] Documentation Update
- [ ] CI Changes
- [ ] Other (please describe):

## Section 2: Draft PR Checklist
- [x] Open draft PR
- [x] Describe the feature that will be added
- [x] Fill out TODO list steps
- [x] Describe requested feedback from reviewers on draft PR
- [ ] Complete Section 8: New Model Checklist (if applicable)

### TODO:
- [x] Implement all three topology checks
- [x] Extend `tech_control_classifiers` to non-SLC models
- [x] Add tests for each new error case
- [x] Update CHANGELOG

### Type of Reviewer Feedback Requested (on Draft PR)
**Implementation feedback**:
Confirm that the three topology checks (length-3 commodity pair detection, storage topology,
and general max-in/out) match the expected behavior described in the issue.

## Section 3: General PR Checklist
- [x] PR description thoroughly describes the new feature, bug fix, etc.
- [x] Added tests for new functionality or bug fixes
- [x] Tests pass (If not, and this is expected, please elaborate in the Section 6: Test Results)
- [x] Documentation
  - [x] Docstrings are up-to-date
  - [ ] Related `docs/` files are up-to-date, or added when necessary
  - [ ] Documentation has been rebuilt successfully
  - [ ] Examples have been updated (if applicable)
- [x] `CHANGELOG.md`
  - [x] At least one complete sentence has been provided to describe the changes made in this PR
  - [ ] After the above, a hyperlink has been provided to the PR using the following format:
    "A complete thought. [PR XYZ]((https://github.com/NatLabRockies/H2Integrate/pull/XYZ)", where
    `XYZ` should be replaced with the actual number.

## Section 4: Related Issues
Solves #809

### Section 5.1: New Files
N/A

### Section 5.2: Modified Files
- `h2integrate/core/h2integrate_model.py`
  - Add `_validate_technology_interconnections` method.
  - Extend `tech_control_classifiers` population to non-SLC models.
  - Call new validation in `setup()`.
- `h2integrate/core/test/test_framework.py`
  - Add `test_validate_technology_interconnections` integration test.
  - Update `test_check_tech_interconnections` second subtest to use a non-storage tech.

## Section 6: Additional Supporting Information

The three checks implemented are:

1. **Discouraged length-3 `[commodity_out, commodity_in]` connections** - When a length-3 connection passes a list pair like `["hydrogen_out", "hydrogen_in"]` where the prefixes match, a `ValueError` is raised prompting the user to use a length-4 connection with an explicit commodity and transport component.

2. **Storage technology topology** - For each storage technology in `tech_control_classifiers`: exactly 1 length-4 input is required; at most 1 length-4 output is allowed; and the direct upstream technology may have at most 2 output streams (one to storage, one to a combiner).

3. **General max-in/out for non-special technologies** - All other technologies connected via length-4 connections (excluding splitters, combiners, storage techs, and the direct upstream of storage techs) must have at most 1 input stream and 1 output stream. The storage upstream exemption applies only to the output-stream count; input streams are still checked.

The fix to always populate `tech_control_classifiers` (removing the `self.slc` guard) ensures checks 2 and 3 work correctly for non-SLC systems where storage is still present. All other usages of `tech_control_classifiers` were already guarded by `if self.slc:` so there are no unintended side effects.

## Section 7: Test Results, if applicable
Validated with targeted test runs on `h2integrate/core/test/test_framework.py` and the full `h2integrate/core/test/` suite. All failures in `test_pose_optimization.py` are pre-existing OpenMDAO API mismatches unrelated to this PR.
- `h2integrate/converters/ammonia/test/test_ammonia_synloop_model.py::test_size_mode_outputs`

All targeted tests passed.

## Section 8 (Optional): New Model Checklist
N/A

### CODE REVIEW

File: h2integrate/core/h2integrate_model.py [hunk 1]
```diff
@@ -2219,6 +2220,126 @@ def _as_commodity_list(commodity):
...
+        # In the technology graph, merged edges (multiple L4 connections between the same pair)
+        # appear as a single edge with a list of commodities, so each unique (source, dest)
+        # pair counts as one stream.
+        in_degs_l4: dict[str, int] = {}
+        out_degs_l4: dict[str, int] = {}
+        for source, dest, commodity in self.technology_graph.edges(data="commodity"):
+            if commodity is None:
+                continue  # length-3 connections carry no commodity; skip them
+            out_degs_l4[source] = out_degs_l4.get(source, 0) + 1
+            in_degs_l4[dest] = in_degs_l4.get(dest, 0) + 1
```
Comment (elenya-grant) [inline h2integrate/core/h2integrate_model.py:2293]: I'm wondering if theres a way to do this with `dict(self.technology_graph.in_degree)` - but then removing non 4-length connections. I can't imagine that it'd actually make the code any simpler - so I like your approach.

File: h2integrate/core/h2integrate_model.py [hunk 2]
```diff
@@ -2219,6 +2220,126 @@ def _as_commodity_list(commodity):
...
+        # --- Check 2: storage technology topology ---
+        storage_techs = [k for k, v in self.tech_control_classifiers.items() if v == "storage"]
+        storage_upstream_techs: set[str] = set()
+        for storage_tech in storage_techs:
+            n_in = in_degs_l4.get(storage_tech, 0)
+            if n_in != 1:
+                raise ValueError(
+                    f"Storage technology {storage_tech!r} has {n_in} input connection(s) in "
+                    f"the technology graph but should have exactly 1."
+                )
```
Comment (elenya-grant) [inline h2integrate/core/h2integrate_model.py:2311]: is there a future case where a storage technology would have 2 different commodity inputs? Like - power and hydrogen? If possible - it'd be nice to check that theres only one input connection per commodity input to the storage.

Comment (elenya-grant) [reply h2integrate/core/h2integrate_model.py:2311]: I think your code would be fine if (for some imaginary storage that takes electricity and hydrogen input) if the electricity and hydrogen come from the same upstream tech. I think that it'd cause an unnecessary error if electricity came from source_a and hydrogen came from source_b.

File: h2integrate/core/h2integrate_model.py [hunk 3]
```diff
@@ -2219,6 +2220,126 @@ def _as_commodity_list(commodity):
...
+                        f"(at most 2 output streams)."
+                    )
+
+        # --- Check 3: max 1 in/out for general technologies connected via length-4 edges ---
+        all_techs_in_l4 = set(in_degs_l4) | set(out_degs_l4)
+        for tech in all_techs_in_l4:
+            if "splitter" in tech or "combiner" in tech:
+                continue
+            if self.tech_control_classifiers.get(tech) == "storage":
+                continue  # already validated in check 2
```
Comment (elenya-grant) [inline h2integrate/core/h2integrate_model.py]: I think we should change the `_control_classifier` for the splitter to be `splitter` and the `_control_classifier` for the combiner to be named `combiner`. This way - we can check the control classifier rather than having logic based on technology names (i.e., a splitter tech has to have "splitter" in the name and a combiner has to have "combiner" in the name)

**Review by elenya-grant — COMMENTED** [review]
really high-level initial review - thanks for addressing this issue! I think it'd be nice to see some tests with some "fake" systems that are more complex than is possible with models in H2I right now to test the limitations of this. For example, looking at systems with converter and storage technologies that have multiple input or output streams. 

Some examples that come to mind are:
- storage tech that has two input commodities (electricity and hydrogen)
- converter that has two input commodities and two output commodities

I also think that the examples in H2I will do a good job of testing this functionality. Thanks John! Will re-review again later (I want to give it a deeper look)

File: h2integrate/core/h2integrate_model.py (hunk 1 above)
Comment (johnjasa) [reply h2integrate/core/h2integrate_model.py:2293]: Thanks for this comment and context! I haven't changed anything here, but good to be aware.

File: h2integrate/core/h2integrate_model.py (hunk 2 above)
Comment (johnjasa) [reply h2integrate/core/h2integrate_model.py:2311]: This is a good point for future cases; I modified the behavior of this check. Now, a storage tech receiving electricity from `source_a` and hydrogen from `source_b` now passes cleanly - two edges but each commodity has exactly one source. The only invalid case is when the same commodity arrives from two different technologies, e.g. hydrogen from both `electrolyzer` and `h2_storage_bypass`, when we should just use a combiner.

File: h2integrate/core/h2integrate_model.py (hunk 3 above)
Comment (johnjasa) [reply h2integrate/core/h2integrate_model.py]: Good call, I changed these two cases:
for `combiner`: "connector" ->"combiner"
for `splitter`: "connector" -> "splitter"

File: examples/36_nuclear_reactor_htse/plant_config.yaml [hunk 4]
```diff
@@ -11,7 +11,7 @@ plant:
 # with the reverse definition.
 # this will naturally grow as we mature the interconnected tech
 technology_interconnections:
-  - [nuclear, htse, [heat_out, heat_in]]
+  - [nuclear, htse, heat, pipe]
```
Comment (elenya-grant) [inline examples/36_nuclear_reactor_htse/plant_config.yaml]: can you use the generic transport model here instead? I think that we should have more examples showing that model because I think it'll be very useful as we develop new models with new commodity streams. I don't think we should add heat as a supposed transport item to pipe.

Comment (elenya-grant) [reply examples/36_nuclear_reactor_htse/plant_config.yaml]: I've meant to make an issue about this - but I was talking to @genevievestarke a while back and we think that we should more examples using the generic transport model - and I think here is a good opportunity to do that.

File: h2integrate/transporters/pipe.py [hunk 5]
```diff
@@ -41,6 +42,8 @@ def setup(self):
             units = "galUS/h"
         elif transport_item == "co2":
             units = "kg/h"
+        elif transport_item == "heat":
+            units = "kW"
```
Comment (elenya-grant) [inline h2integrate/transporters/pipe.py]: see my comment earlier for example 36. I don't think we should add heat to the pipe right now.

**Review by elenya-grant — COMMENTED** [review]
just left some small comments

File: h2integrate/core/h2integrate_model.py (hunk 2 above)
Comment (elenya-grant) [reply h2integrate/core/h2integrate_model.py:2311]: I agree - that sounds great! Thank you! I think thats perfect.

Comment (elenya-grant) [reply h2integrate/core/h2integrate_model.py:2311]: In PR #816, I added a fake system that handles another type of complex system I imagine we could have in H2I. I don't think that you have to use that specific fake system - but could be good to use something like that to test these checks for the connections into/out of a converter tech (h2integrate/control/control_strategies/system_level/test/test_slc_baseclass.py::test_find_converter_techs_fake_system)

File: h2integrate/core/h2integrate_model.py [hunk 6]
```diff
@@ -2219,6 +2220,159 @@ def _as_commodity_list(commodity):
...
+        for connection in technology_interconnections:
+            if len(connection) != 3:
+                continue
+            connected_parameter = connection[2]
+            if not isinstance(connected_parameter, list | tuple) or len(connected_parameter) != 2:
+                continue
+            source_param, dest_param = connected_parameter
+            if not isinstance(source_param, str) or not isinstance(dest_param, str):
+                continue
+            if source_param.endswith("_out") and dest_param.endswith("_in"):
```
Comment (elenya-grant) [inline h2integrate/core/h2integrate_model.py]: Not trying to nit-pick but I just remembered PR #774 and wonder if we should do:

```python
if source_param.split("[")[0].endswith("_out") and dest_param.split("[")[0].endswith("_in"):
```

File: h2integrate/core/h2integrate_model.py [hunk 7]
```diff
@@ -2219,6 +2220,159 @@ def _as_commodity_list(commodity):
...
+                if commodity_from_source == commodity_from_dest:
+                    source_tech, dest_tech = connection[0], connection[1]
+                    raise ValueError(
+                        f"Connection [{source_tech!r}, {dest_tech!r}, "
+                        f"[{source_param!r}, {dest_param!r}]] passes commodity "
+                        f"{commodity_from_source!r} between technologies using a "
+                        f"length-3 format. Use a length-4 connection instead: "
+                        f"[{source_tech!r}, {dest_tech!r}, {commodity_from_source!r}, "
+                        f"'<transport_tech>']. Use the generic transport component if "
+                        f"no specific transport is needed."
```
Comment (elenya-grant) [inline h2integrate/core/h2integrate_model.py]: Could we rephrase this part of the error message: "Use the generic transport component if no specific transport is needed"?

I think that the generic transport component is useful if a) no specific transport is needed and b) it's a commodity that isn't supported by cable or pipe models (perhaps because its a new commodity). I think the most simple re-phrasing would be "You can use the `GenericTransporterPerformanceModel` to transport {commodity_from_source!r}"? But I'm open to other ideas

File: h2integrate/core/test/test_framework.py [hunk 8]
```diff
@@ -150,11 +153,306 @@ def test_check_tech_interconnections(subtests, temp_copy_of_example):
...
+        ["source_h2", "storage", "hydrogen", "pipe"],
+        ["storage", "h2_combiner", "hydrogen", "pipe"],
+    ]
+    classifiers = {
+        "storage": "storage",
+        "h2_combiner": "combiner",
+    }
+    fake = _make_fake_model(interconnections, classifiers)
+
+    with subtests.test("two-commodity storage passes validation"):
```
Comment (elenya-grant) [inline h2integrate/core/test/test_framework.py:307]: I know that this use-case is still not something we imagine happening a ton yet - but in my head, this should be a valid system of interconnections but an error is thrown:

```python
interconnections = [
        ["source_elec", "storage", "electricity", "cable"],
        ["source_h2", "storage", "hydrogen", "pipe"],
        ["storage", "h2_combiner", "hydrogen", "pipe"],
        ["source_h2", "h2_combiner", "hydrogen", "pipe"],
        ["source_elec", "elec_combiner", "electricity", "cable"],
        ["storage", "elec_combiner", "electricity", "cable"],
]
classifiers = {
        "storage": "storage",
        "h2_combiner": "combiner",
        "elec_combiner": "combiner",
}
```

File: h2integrate/core/h2integrate_model.py [hunk 9]
```diff
@@ -2219,6 +2220,159 @@ def _as_commodity_list(commodity):
...
+            for upstream_tech in upstream_techs_l4:
+                storage_upstream_techs.add(upstream_tech)
+                n_out_upstream = out_degs_l4.get(upstream_tech, 0)
+                if n_out_upstream > 2:
+                    raise ValueError(
+                        f"Technology {upstream_tech!r} feeds storage technology "
+                        f"{storage_tech!r} but has {n_out_upstream} output connection(s). "
+                        f"It should connect only to {storage_tech!r} and a combiner "
+                        f"(at most 2 output streams)."
+                    )
```
Comment (elenya-grant) [inline h2integrate/core/h2integrate_model.py]: I think that here we should check that n_out_upstream based on the commodity

```suggestion
            for upstream_tech in upstream_techs_l4:
                storage_upstream_techs.add(upstream_tech)
                for commodity in self.technology_graph.edges[upstream_tech, storage_tech].get("commodity"):
                
                n_out_upstream = out_commodity_dests.get(upstream_tech, {}).get(commodity,0)
                if n_out_upstream > 2:
                    raise ValueError(
                        f"Technology {upstream_tech!r} feeds storage technology "
                        f"{storage_tech!r} but has {n_out_upstream} output connection(s). "
                        f"It should connect only to {storage_tech!r} and a combiner "
                        f"(at most 2 output streams)."
                    )
```

File: h2integrate/core/h2integrate_model.py [hunk 10]
```diff
@@ -2219,6 +2220,159 @@ def _as_commodity_list(commodity):
...
+                        f"Storage technology {storage_tech!r} receives commodity "
+                        f"{commodity!r} from {n_sources} sources in the technology graph "
+                        f"but should receive it from at most 1."
+                    )
+            n_out = out_degs_l4.get(storage_tech, 0)
+            if n_out > 1:
+                raise ValueError(
+                    f"Storage technology {storage_tech!r} has {n_out} output connection(s) in "
+                    f"the technology graph but should have at most 1."
+                )
```
Comment (elenya-grant) [inline h2integrate/core/h2integrate_model.py]:
```suggestion
            for commodity, n_out in out_commodity_dests.get(storage_tech, {}).items():
                if n_out > 1:
                    raise ValueError(
                        f"Storage technology {storage_tech!r} has {n_out} output connection(s) in "
                        f"of commodity {commodity} but should have at most 1."
                    )
```

**Review by elenya-grant — CHANGES_REQUESTED** [review]
Just left some comments on some edge cases that I think are valid but would cause errors when they shouldn't. I can share the tests I made to check these if you want. I will finish this review tommorrow! Looks good so far! Only some unlikely (at the moment) edge-cases that I found that this would throw unnecessary errors for

File: h2integrate/core/test/test_framework.py (hunk 8 above)
Comment (johnjasa) [reply h2integrate/core/test/test_framework.py:307]: Thanks for the example connections. The main difference here is that a singular storage technology accepts and stores both hydrogen and electricity, is that right? I don't immediately know of a physical system that behaves like that -- do you? I would argue we don't need to support this edge case at this moment, but let me know if you have a need or example case for it.

Comment (elenya-grant) [reply h2integrate/core/test/test_framework.py:307]: you're right - I can't imagine a physical system that behaves like this. Change isn't needed!

File: h2integrate/core/test/test_framework.py [hunk 11]
```diff
@@ -150,11 +153,324 @@ def test_check_tech_interconnections(subtests, temp_copy_of_example):
...
+        ["converter", "dest_b", "ammonia", "pipe"],  # ammonia goes to two places
+    ]
+    bad = _make_fake_model(bad_interconnections, {"converter": "dispatchable"})
+
+    with subtests.test("same output commodity to two destinations raises error"):
+        with pytest.raises(ValueError) as excinfo:
+            H2IntegrateModel._validate_technology_interconnections(bad)
+        err = str(excinfo.value)
+        assert "converter" in err
+        assert "destinations" in err
```
Comment (elenya-grant) [inline h2integrate/core/test/test_framework.py]: could we have more detail in checking these error messages? I like there to be enough info here that someone could cntrl+f and find the error message in H2I.

File: h2integrate/core/test/test_framework.py [hunk 12]
```diff
@@ -150,11 +153,324 @@ def test_check_tech_interconnections(subtests, temp_copy_of_example):
...
+        assert "length-4 connection" in err
+
+
+@pytest.mark.unit
+def test_validate_interconnections_length3_commodity_pair_with_slices_raises(subtests):
+    """A length-3 [X_out[...], X_in[...]] parameter pair must raise the
+    same guidance to use a length-4 commodity connection."""
+    interconnections = [
+        ["wind", "electrolyzer", ["electricity_out[0:4]", "electricity_in[0:4]"]],
+    ]
```
Comment (elenya-grant) [inline h2integrate/core/test/test_framework.py:453]: thanks for adding this test in!

**Review by elenya-grant — COMMENTED** [review]
Thanks for making these changes John! I haven't dug deeply into the logic introduced to handle splices (I can review it in more detail if you'd like). My only remaining suggestion is that I'd like to see more info on the strings for the error message assert statements. As-is, I think it'd be hard to debug a test failure and to be really confident the expected error message is being raised.

I imagine that as H2I develops, we will run into some use-cases that test the bounds of this logic but I think this is great for now!

File: h2integrate/core/test/test_framework.py (hunk 11 above)
Comment (johnjasa) [reply h2integrate/core/test/test_framework.py]: Thanks for the suggestion! I added more words to the asserts in the tests

**Review by elenya-grant — APPROVED** [review]
Looks good to me! thanks John!
